### PR TITLE
shutdown: Add Xen kexec support

### PR DESCRIPTION
**Pull request by sthibaul**
In the Xen case, it's the hypervisor which manages kexec. We thus have to ask it through libxenctrl whether a kernel is loaded, instead of relying on /sys/kernel/kexec_loaded.

### CODE REVIEW

File: src/basic/util.c
Comment (sthibaul) [reply src/basic/util.c]: I have reworked the source code a bit to avoid the warning.

File: src/basic/util.c
Comment (bluca) [inline src/basic/util.c]: `r`, and no need to initialise here

File: src/basic/util.c
Comment (bluca) [inline src/basic/util.c]: move the ifdefs inside the function, with an `#else` that returns -EOPNOTSUPP

File: src/basic/util.c
Comment (bluca) [inline src/basic/util.c]: `return log_debug_errno(errno, "Cannot access /proc/xen: %m)`

File: src/basic/util.c
Comment (bluca) [inline src/basic/util.c]: `return log_debug_errno(errno, "Failed to open xen control interface: %m)` (assuming errno is set)

File: src/basic/util.c
Comment (bluca) [inline src/basic/util.c]: `int r` and initialize it separately from the declaration as per coding style

**Review by bluca — CHANGES_REQUESTED** [review]
You'll probably want to add libxenctrl to various distro configs in mkosi.default.d/*

Also, given this is in src/basic, we probably need to do the dlopen dance given it's on auto? Otherwise every binary will get this dependency

**Comment by sthibaul** [comment]
> Also, given this is in src/basic, we probably need to do the dlopen dance given it's on auto? Otherwise every binary will get this dependency

I understand the dlopen dance, but that's in `shared`, and there doesn't seem to be a defined way to make files from `basic` use functions from `shared`?

Perhaps we should move the `kexec_loaded` function into `shared`? It's apparently only used in `initctl`, `logind-dbus`, and `systemctl*`.

**Comment by bluca** [comment]
> > Also, given this is in src/basic, we probably need to do the dlopen dance given it's on auto? Otherwise every binary will get this dependency
> 
> I understand the dlopen dance, but that's in `shared`, and there doesn't seem to be a defined way to make files from `basic` use functions from `shared`?
> 
> Perhaps we should move the `kexec_loaded` function into `shared`? It's apparently only used in `initctl`, `logind-dbus`, and `systemctl*`.

Yes, moving to shared makes sense, please do - thanks!

File: mkosi.default.d/centos_epel/10-mkosi.centos_epel
Comment (daandemeyer) [inline mkosi.default.d/centos_epel/10-mkosi.centos_epel]: Please also install the runtime package in `Packages=` for all distros

File: src/shared/xen-util.c
Comment (daandemeyer) [inline src/shared/xen-util.c]: Please use `FOREACH_STRING`

File: src/shared/xen-util.h
Comment (daandemeyer) [inline src/shared/xen-util.h]: Please implement this as we do for other dlopen() functions (put the ifdef in the function implementation instead of in the header)

File: src/shared/xen-util.h
Comment (daandemeyer) [inline src/shared/xen-util.h]: Please drop the argument names to be consistent with the other dlopen stuff

**Review by bluca — APPROVED** [review]

**Comment by sthibaul** [comment]
(I have just dropped the xen deps in the centos case, apparently they have dropped those packages in versions 8 & 9)

**Comment by sthibaul** [comment]
(fixed the libxen-dev package name on debian/ubuntu)

File: src/shared/reboot-util.c
Comment (yuwata) [inline src/shared/reboot-util.c]: `xen_kexec_loaded()` is a boolean function with negative errno. Hence, in this case, please simply use `return r;`.

File: meson.build
Comment (yuwata) [inline meson.build]: We usually align with `(`.

**Review by yuwata — APPROVED** [review]
LGTM. Several minor nits.

File: meson.build
Comment (medhefgo) [inline meson.build]: Please use meson features for new options added in. See https://github.com/systemd/systemd/pull/24175 for how this is supposed to be done.

**Comment by poettering** [comment]
grmbl. Is there no way we can make this work without the library dep? is there no flag file or so in the file system that can tell us abou this? what precesily does this xen library actually do internally, when checking this?

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: hmm, are you sure xc_interface_open() sets errno? I can't find any docs, and google for code snippets shows noone using errno afterwards?

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: hmm, what does this return? I presume that if also returns errors in errno, no? 

hence, shouldn't this be:

```c
r = RET_NERRNO(sym_xc_kexec_status(xch, KEXEC_TYPE_DEFAULT));
```

(under the assumtpion that these calls communicate errors via errno, as your xc_interface_open() call above appears to suggests)

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c:193]: i would very much prefer if we would put this check first, as it is the more common case.

Comment (poettering) [reply src/shared/reboot-util.c:193]: also, if you touch this, we nowadays prefer if unexpected errors are debug logged. i.e. ENOENT we kinda expect, any other errors we don't. hence, please change this to:

```c
r = read_one_line_file("/sys/kernel/kexec_loaded", &s);
if (r < 0) {
        if (r != -ENOENT) 
                log_debug_errno(r, "Unable to read /sys/kernel/kexec_loaded, ignoring: %m");
        return false;
}
if (s[0] == '1'))
        return true;

return xen_kexec_loaded();
```

File: src/shared/xen-util.c
Comment (poettering) [inline src/shared/xen-util.c]: wow, this library is happy to break ABIs!

File: src/shared/xen-util.c
Comment (poettering) [inline src/shared/xen-util.c]: should be at debug level

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: this check is much more light-weight than the dlopen_xenctrl() check, hence please do this first.

File: meson.build
Comment (bluca) [reply meson.build]: It's too soon for that, it breaks backward compatibility, so for now this is fine as-is

File: src/shared/reboot-util.c
Comment (bluca) [reply src/shared/reboot-util.c]: xen's code itself assumes it does in a few places:

https://sources.debian.org/src/xen/4.16.2-2/tools/libs/light/libxl_save_helper.c/?hl=223#L223
https://sources.debian.org/src/xen/4.16.2-2/tools/python/xen/lowlevel/xc/xc.c/?hl=69#L69

but it doesn't look like the implementation does:

https://sources.debian.org/src/xen/4.16.2-2/tools/libs/ctrl/xc_private.c/?hl=27#L74

which is strange... anyway, please then use `return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Failed to open Xen control interface.");`

**Comment by bluca** [comment]
> grmbl. Is there no way we can make this work without the library dep? is there no flag file or so in the file system that can tell us abou this? what precesily does this xen library actually do internally, when checking this?

Probably IPC to the hypervisor? It's optional, only used on-demand and it's a tiny snippet so I don't see any issues with using it, it's as non-invasive as it could be

File: meson.build
Comment (medhefgo) [reply meson.build]: It does not. meson features are fine for *new* options we add as we don't have to adhere to compat for those. The minimum meson version we use already supports features. It's only the compat mapping of old/custom options to features which requires newer meson version.

So I'd rather we use meson stuff as intended where it makes sense. And a new option doesn't have to adhere to old ways of doing things.

**Comment by medhefgo** [comment]
Dear @poettering, can you please learn to use the PR review feature of github?

It's as simple as clicking on the green "Start Review" button instead of the "Add single comment". It allows you to add as many comments throughout all the commits/files in a single PR in one go. You can even delete/revise them if you have the need to later during the review. When you're done you just use the green "Review Changes" button on the top right. You don't even have to commit to a approval/requesting changes as you can simply select "Comment" in the drop down (which is the default).

Why should you do this, you may ask? Because if you use single comments, they are slowly trickled out, spamming everyone's mailbox, which is quite annoying. It also blows up the mail threads needlessly. The review features sends out all the comments in one go. It also serves as a nice sequence point of "is he done reviewing by now or not".

Thanks

**Comment by poettering** [comment]
@medhefgo so the reason i don't use the review stuff always is that it delays the posting of it. For stuff that already has a green label that sucks, because noone can see that there will be a bunch of comments coming in as a review. Quite often I do reviews over a long period of time, and it really sucks if people end up merging PRs under my feet for which i have a half-complete review. Sure, I could drop the green label first, but that would appear quite weird, if I drop the label and noone has any idea why, because the review is still gonna take some hours more.

And it's actually worse: if people force push a new version onto a PR then any pending review that hasn't been posted is a total mess. And puts the pressure on the review to quickly push comments out so that one doesn't run inhearto this mess.

github UI is just bad sometimes, this is one of these cases. I do reviews *all* *the* *time* and I ran into this over and over again, so I tweaked my workflow to avoid this.

**Comment by sthibaul** [comment]
> grmbl. Is there no way we can make this work without the library dep?

Unless we copy/paste the whole hypercall machinery from libxc, we can't.

> is there no flag file or so in the file system that can tell us abou this?

It's not as simple as a flag file in the file system, no.

> what precesily does this xen library actually do internally, when checking this?

It performs a hypercall into the hypervisor.

File: src/shared/reboot-util.c
Comment (sthibaul) [reply src/shared/reboot-util.c]: > but it doesn't look like the implementation does:
>
> https://sources.debian.org/src/xen/4.16.2-2/tools/libs/ctrl/xc_private.c/?hl=27#L74

In the err cases, errno has been set by previous functions calls.

File: src/shared/reboot-util.c
Comment (sthibaul) [reply src/shared/reboot-util.c]: > hmm, what does this return? I presume that if also returns errors in errno, no?

No, in this case it returns the error as negative value.

File: src/shared/reboot-util.c
Comment (sthibaul) [reply src/shared/reboot-util.c:193]: > i would very much prefer if we would put this check first, as it is the more common case.

No: the non-existence of `/sys/kernel/kexec_loaded` doesn't say anything about whether the hypervisor has loaded a kexec kernel. And even if a user somehow loads a kernel by hand into linux (and thus `/sys/kernel/kexec_loaded` contain 1), kexec-ing with linux would just fail, it's really the hypervisor that needs to be contacted at all, when running under Xen.

File: src/shared/xen-util.c
Comment (bluca) [reply src/shared/xen-util.c]: those are not ABI breaks, looks like a build/packaging issue - instead of shipping the full library as `libxenctrl.so.4.14.0` and the symlink as `libxenctrl.so.4`, the symlink includes the ABI revision as well, for some reason, which gets bumped on every new symbol

nothing we can do about it here, except maybe look for `libxenctrl.so.4` first in case it ever gets added?

File: meson.build
Comment (sthibaul) [reply meson.build]: Could you get to agree? :)

File: src/shared/xen-util.c
Comment (bluca) [reply src/shared/xen-util.c]: @sthibaul do you think there's any chance this will ever get solved? I'd add it regardless, it's cheap enough in this context to look for one extra file

Comment (sthibaul) [reply src/shared/xen-util.c]: It's not "bumped on every new symbol", it's simply the version of Xen.

I don't think they will ever add a `.4` symlink: they do break ABI every now and then. Not for simple calls like `xc_kexec_status`, but for more involved calls like creating a domain, yes.

File: meson.build
Comment (bluca) [reply meson.build]: leave as-is, an unrelated PR is not the right place to bikeshed on the style of the meson files

File: src/shared/xen-util.c
Comment (bluca) [reply src/shared/xen-util.c]: Any chance you could ask? As you can see here, it's going to be a massive pain to keep this list up to date, and none of us use Xen anywhere so I'd be up to you to send PRs to update it every time they release a new version...

Comment (daandemeyer) [reply src/shared/xen-util.c]: Should we be smarter about finding the SO? There's nothing preventing us from figuring out the installed SO in meson and setting a define and looking that up at runtime.

File: meson.build
Comment (medhefgo) [reply meson.build]: > leave as-is, an unrelated PR is not the right place to bikeshed on the style of the meson files

The fuck? This introduces a new feature and thus should do this properly. Old cold being old code is fine. But new code should adhere and use new paradigms of the language. meson features are far superior for this as they make dealing with dependencies far easier, with less code.

File: src/shared/xen-util.c
Comment (sthibaul) [reply src/shared/xen-util.c]: One still has to check on each release that the ABI hasn't been changed. Even if that's little probable, it may happen.

Comment (sthibaul) [reply src/shared/xen-util.c]: > Any chance you could ask?

Really, there is no chance that they add such a link. They really don't promise any ABI stability for xenctrl.

Comment (bluca) [reply src/shared/xen-util.c]: unfortunate, but I guess there's not much that can be done about it - please do keep in mind that none of the maintainers is in a position to keep this up to date, as afaik nobody here uses xen, so we'll rely on xen users like yourself to send updates

File: meson.build
Comment (bluca) [reply meson.build]: Even if it is, and I'm not sure that's the case as having 2 ways to do the same thing might get confusing and might not be worth it, again a PR fixing an actual bug on a platform from a first-time contributor is _not_ the right place to bikeshed on something like this. Instead please open a follow-up PR after this is merged, so that it can be discussed without involving @sthibaul who has no skin in this particular game and should not be asked to spend time on it.

**Comment by sthibaul** [comment]
> > grmbl. Is there no way we can make this work without the library dep?
> 
> Unless we copy/paste the whole hypercall machinery from libxc, we can't.

Here is how it looks like:

```
#include <fcntl.h>
#include <error.h>
#include <errno.h>
#include <unistd.h>
#include <assert.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/ioctl.h>

#define __XEN_INTERFACE_VERSION__ 0x00040900
#include <xen/xen.h>
#include <xen/kexec.h>
#include <xen/sys/privcmd.h>

int main(void) {
	int privcmd_fd = open("/dev/xen/privcmd", O_RDWR|O_CLOEXEC);
	if (privcmd_fd == -1)
		error(1, errno, "open privcmd");

	int buf_fd = open("/dev/xen/hypercall", O_RDWR|O_CLOEXEC);
	if (buf_fd == -1)
		error(1, errno, "open buf_fd");

	size_t page_size = sysconf(_SC_PAGE_SIZE);
	void *buffer = mmap(NULL, page_size, PROT_READ|PROT_WRITE, MAP_SHARED, buf_fd, 0);
	if (buffer == MAP_FAILED)
		error(1, errno, "allocate buffer");
	memset(buffer, 0, page_size);

	xen_kexec_status_t *status = buffer;
	assert(sizeof(*status) <= page_size);
	status->type = KEXEC_TYPE_DEFAULT;

	privcmd_hypercall_t call = {
		.op = __HYPERVISOR_kexec_op,
		.arg = { KEXEC_CMD_kexec_status, (__u64)(uintptr_t) &status },
	};
	int ret = ioctl(privcmd_fd, IOCTL_PRIVCMD_HYPERCALL, &call);
	printf("%d\n", ret);

	return 0;
}
```

It's not that bad. Perhaps we can go that way, rather than fiddling with libxenctrl versions?

**Comment by sthibaul** [comment]
> > > grmbl. Is there no way we can make this work without the library dep?
> > 
> > 
> > Unless we copy/paste the whole hypercall machinery from libxc, we can't.
> 
> Here is how it looks like:

(and yes, *that* interface is cast into stone, Xen won't change it ever)

**Comment by sthibaul** [comment]
> So we only need the library for the headers?

Yes.

File: meson.build
Comment (medhefgo) [reply meson.build]: This isn't bikeshedding. It is requesting to use the *proper meson interface* do add a *new library* instead of using the custom ad-hoc method practiced until then. And it's not a hard burden to do it either, it literally makes the meson code smaller and idiomatic!

As I see it, we practice this all the time: old code can stay using old/deprecated interfaces/processes, but we expect new code to adhere to the new/proper/idiomatic way things are done. Why should our build system be exempt here?

Comment (medhefgo) [reply meson.build]: Btw, this is what I was asking for (see how it's smaller, easier to follow, idiomatic and no burden to use?):
```meson
libxenctrl = dependency('xencontrol',
                      version : '>= 4.9',
                      required : get_option('xencontrol'))
conf.set10('HAVE_XENCTRL', libxenctrl.found())
```

```
option('xenctrl', type : 'feature',
       description : 'support for Xen kexec')
```

**Comment by sthibaul** [comment]
Here is the new version.

File: src/shared/reboot-util.c
Comment (bluca) [inline src/shared/reboot-util.c]: Sort alphabetically within this chunk

File: src/shared/reboot-util.c
Comment (bluca) [inline src/shared/reboot-util.c]: Move this block above, below the other system headers

File: src/shared/reboot-util.c
Comment (bluca) [inline src/shared/reboot-util.c]: status is not initialized here, maybe use the type instead? It looks a bit weird

Comment (medhefgo) [reply src/shared/reboot-util.c]: `sizeof` only evaluates the type. This form is usually preferred as it remains valid even if the type of a var changes.

File: src/shared/reboot-util.c
Comment (bluca) [inline src/shared/reboot-util.c:134]: given this will run on every kexec, we'll always get at least one spurious debug log message - maybe it's worth to do a `if (access("/proc/xen", F_OK) < 0) return -EOPNOTSUPP` as the very first thing? Without logging the result. That way there's no extra noise in the "normal" case?

File: src/shared/reboot-util.c
Comment (bluca) [reply src/shared/reboot-util.c]: I know, it just looks a bit strange at a glance, but it's not important

File: src/shared/reboot-util.c
Comment (bluca) [inline src/shared/reboot-util.c]: duplicated

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: merge these two lines please

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: duplicate ;;

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: this cannot really happen, can it? this would be a programming error, not a realistic error no? hence maybe make this an assert()?

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: we usually check fd validity with `if (privcmd_fd < 0)`...

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: ditto

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: memzero(buffer, size)

File: src/shared/reboot-util.c
Comment (sthibaul) [reply src/shared/reboot-util.c]: I doubt we'd ever see a system where a page size would be less than the kexec request structure, but I'd still rather see a nice error in case page_size() misbehaves, than an assert.

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: instead of the memzero() and the explicit field initialization here, maybe also use compound initialization here, i.e.:

```c
*(*xen_kexec_status_t) buffer = (xen_kexec_status_t) {
        .type = KEXEC_TYPE_DEFAULT,
};
```
(of course, this would initialize less than the whole page, most likely, but that should be ok, no?)

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: maybe better:

```c
privcmd_hypercall_t call = {
        .op = __HYPERVISOR_kexec_op,
        .arg = { 
                KEXEC_CMD_kexec_status, 
                PTR_TO_UINT64(&status),
        },
};
```

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: pass `r` as first argument to log_debug_errno()

**Comment by poettering** [comment]
ah, the ioctl() call is really nice. I like.

File: src/shared/reboot-util.c
Comment (poettering) [reply src/shared/reboot-util.c:193]: please make the log_debug_errno() change as mentioned, i.e. if you touch this, please adjust this to our current coding style.

(this is independent on the order in which you do the two questions)

**Review by bluca — APPROVED** [review]

File: src/shared/meson.build
Comment (bluca) [inline src/shared/meson.build:469]: is this still needed, given it's no longer linking?

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: please change this to:

```c
if (access("/proc/xen", F_OK) < 0) {
        if (errno == ENOENT)
                return -EOPNOTSUPP;
        return log_debug_errno(errno, "Unable to test whether /proc/xen exists: %m");
}
```

File: src/shared/reboot-util.c
Comment (poettering) [inline src/shared/reboot-util.c]: downgrade to log_debug_errno() please

**Review by poettering — CHANGES_REQUESTED** [review]
looks pretty good

File: src/shared/meson.build
Comment (sthibaul) [reply src/shared/meson.build:469]: It'd be needed for getting the `-I` flags to access the headers.

File: src/shared/reboot-util.c
Comment (sthibaul) [reply src/shared/reboot-util.c]: Why? this is a sign that the system is behaving oddly, it's better making sure it gets logged to explain why things failed.

Comment (poettering) [reply src/shared/reboot-util.c]: we have this rule in systemd: functions either log about all their errors or about none (in which case the caller may decide to log). Whenever you write a function, figure out which kind of function yours is: a logging one (typically the more main-program-like stuff) or a non-logging one (typically library-like stuff). 

Via this rule we ensure that every error gets logged exactly once, and not accidentally eaten up entirely, or logged about twice (both of which are confusing to the user).

(note that LOG_DEBUG logging does not count as "logging" in this context, debug logging is always OK.)

https://systemd.io/CODING_STYLE/#error-handling

Your function so far looks like a non-logging one (and I think that's the right choice, given this is a library-like thing). Hence it should not log above LOG_DEBUG, on any code path.

**Comment by poettering** [comment]
looks good, besides the one log level thing

**Comment by poettering** [comment]
lgtm. thank you very much for reworking this based on the direct ioctl() calling, this is much nicer.

**Pull request merged**
